README.md
# Identification of Languages and Language Groups for OCaml

## Synopsis

This library defines abstract types for languages and language families
covered by ISO 639 and provides related functionality and conversions.

- `Lang.t` maps one-to-one to ISO 639-3 language codes, which covers
  individual languages and macrolanguages.
- `Lang_family.t` maps one-to-one to ISO 639-5 language codes, which covers
  language families and groups.
- `Lang_or_family.t` is the sum of the two above types.

## Example

```ocaml
utop # #require "iso639";;
utop # open Iso639;;
utop # #install_printer Lang.pp;;

utop # let por = Lang.of_string_exn "por";;
val por : Lang.t = por
utop # Lang.to_iso639p1 por;;
- : string option = Some "pt"
utop # Lang.macrolanguage por;;
- : Lang.t option = None
utop # Lang.macrolanguage_members por;;
- : Lang.t list = []

utop # let nor = Lang.of_string_exn "nor";;
val nor : Lang.t = nor
utop # Lang.macrolanguage_members nor;;
- : Lang.t list = [nob; nno]
utop # Lang.macrolanguage (Lang.of_string_exn "nob");;
- : Lang.t option = Some nor
```

## Limitations

Only part of ISO 639 which are expected to be commonly used by applications
are implemented.

The library does not include the English end French names of languages, and
there is no plan to add them, as this would significantly increase the size
of the library.


CHANGES.md
## v0.0.3 - 2019-03-24

- Keep track of external data files using a Git-tracked manifest.
- Update data.

## v0.0.2 - 2018-11-28

- Split API into three modules corresponding to ISO 639-3, ISO 639-5 and
  combined types.
- Add scope and macrolanguage requests.
- Tidy up the API and documentation.
- Bugs may have been fixed in the process.

## v0.0.1 - 2018-09-04

- Initial release.


tests/test_iso639.ml
(* Copyright (C) 2018  Petter A. Urkedal <paurkedal@gmail.com>
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version, with the OCaml static compilation exception.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *)

open Iso639
open Printf

let check_lang lang =
  assert (Lang.equal (Lang.of_int_exn (Lang.to_int lang)) lang);
  (match Lang.of_string (Lang.to_string lang) with
   | None -> assert false
   | Some lang' -> assert (Lang.equal lang lang'))

let check_lang_family lang =
  assert (Lang_family.equal (Lang_family.of_int_exn (Lang_family.to_int lang))
                            lang)

let check_alpha2 p1_count alpha2 =
  (match Lang_or_family.of_part1_string alpha2 with
   | Some lang ->
      incr p1_count;
      assert (Lang_or_family.to_part1_string lang = Some alpha2)
   | None -> ())

let check_alpha3 p2_count p3_count p5_count alpha3 =
  let langI = Lang.of_string alpha3 in
  let lang3 = Lang_or_family.of_part3_string alpha3 in
  let langF = Lang_family.of_string alpha3 in
  let lang5 = Lang_or_family.of_part5_string alpha3 in
  (match langI, lang3, langF, lang5 with
   | None, None, None ,None -> ()
   | Some langI, Some lang3, None, None ->
      incr p3_count;
      assert (Lang_or_family.to_part3_string lang3 = Some alpha3);
      assert (Lang_or_family.scope lang3 <> `Collective);
      assert (Lang_or_family.equal (Lang.to_lang_or_family langI) lang3);
      check_lang langI
   | None, None, Some langF, Some lang5 ->
      incr p5_count;
      assert (Lang_or_family.to_part5_string lang5 = Some alpha3);
      assert (Lang_or_family.scope lang5 = `Collective);
      assert (Lang_or_family.equal (Lang_family.to_lang_or_family langF) lang5);
      check_lang_family langF
   | _ -> assert false);
  (match Lang_or_family.of_part2_string alpha3 with
   | Some lang2 ->
      let alpha_p2t = Lang_or_family.to_part2t_string lang2 in
      let alpha_p2b = Lang_or_family.to_part2b_string lang2 in
      let is_bib =
        (match alpha_p2t, alpha_p2b with
         | None, None -> assert false
         | None, Some _ | Some _, None -> assert false
         | Some p2t, Some p2b ->
            incr p2_count;
            assert (p2t = alpha3 || p2b = alpha3);
            alpha3 <> p2t) in
      (match lang3, lang5 with
       | None, None -> assert is_bib
       | Some langI, None -> assert (Lang_or_family.equal lang2 langI)
       | None, Some langF -> assert (Lang_or_family.equal lang2 langF)
       | Some langI, Some langF ->
          assert (not (Lang_or_family.equal langI langF)))
   | None -> ())

let check_scope () =
  let chk langI scope =
    let lang =
      (match Lang_or_family.of_part3_string langI with
       | Some lang -> lang
       | None -> assert false) in
    assert (Lang_or_family.scope lang = scope) in
  chk "spa" `Individual;
  chk "nor" `Macro;
  chk "mis" `Special

let () =
  let p1_count = ref 0 in
  let p2_count = ref 0 in
  let p3_count = ref 0 in
  let p5_count = ref 0 in
  for i0 = 0 to 255 do
    let ch0 = Char.chr i0 in
    for i1 = 0 to 255 do
      let ch1 = Char.chr i1 in
      let alpha2 = sprintf "%c%c" ch0 ch1 in
      check_alpha2 p1_count alpha2;
      for i2 = 0 to 255 do
        let ch2 = Char.chr i2 in
        let alpha3 = sprintf "%c%c%c" ch0 ch1 ch2 in
        check_alpha3 p2_count p3_count p5_count alpha3
      done
    done
  done;
  check_scope ()


lib/lang.ml
(* Copyright (C) 2018  Petter A. Urkedal <paurkedal@gmail.com>
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version, with the OCaml static compilation exception.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *)

include Common

let of_int lang =
  if is_valid_part3 lang then Some lang else None

let of_int_exn lang =
  if is_valid_part3 lang then lang else failwith "Iso639.Lang.of_int_exn"

let of_lang_or_family lang =
  let i = Lang_or_family.to_int lang in
  if i < 0x8000 then Some i else None

let to_lang_or_family = Lang_or_family.of_int_unsafe

let of_string = of_iso639p3

let of_string_exn s =
  (match of_iso639p3 s with
   | None -> invalid_arg "Iso639.Lang.of_string_exn"
   | Some lang -> lang)

let scope = Data.lang3_scope

let macrolanguage lang =
  let langM = Data.lang3_macrolanguage lang in
  if langM = lang then None else Some langM

let macrolanguage_members lang =
  let s = Data.lang3_macrolanguage_members lang in
  List.init (String.length s / 2)
    (fun i -> Char.code s.[2*i] lsl 8 lor Char.code s.[2*i + 1])


lib/common.ml
(* Copyright (C) 2018  Petter A. Urkedal <paurkedal@gmail.com>
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version, with the OCaml static compilation exception.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *)

type t = int
let equal = (=)
let compare = compare
let to_int x = x

let is_valid_part3 lang =
  lang < 0x8000 &&
  Char.code (Data.is_iso639p3_bits.[lang / 8]) lsr (lang mod 8) land 1 = 1

let is_part1 x = Data.to_iso639p1 x <> x
let is_part2 x = Data.is_iso639p2t x

let alpha_of_int x = Char.chr (x mod 32 + 0x60)

let int_of_alpha = function
 | 'a'..'z' as c -> Char.code c - 0x60
 | _ -> raise Not_found

let alpha2_of_int x = String.init 2 (fun i -> alpha_of_int (x lsr (5*(1 - i))))
let alpha3_of_int x = String.init 3 (fun i -> alpha_of_int (x lsr (5*(2 - i))))

let int_of_alpha3 s =
  if String.length s <> 3 then raise Not_found else
  int_of_alpha s.[0] lsl 10 + int_of_alpha s.[1] lsl 5 + int_of_alpha s.[2]

let to_string = alpha3_of_int

let pp ppf lang = Format.pp_print_string ppf (to_string lang)

let to_part1_string lang =
  let lang1 = Data.to_iso639p1 lang in
  if lang1 = lang then None else Some (alpha2_of_int lang1)

let of_part1_string s =
  if String.length s <> 2 then None else
  try
    let lang1 = int_of_alpha s.[0] lsl 5 + int_of_alpha s.[1] in
    let lang3 = Data.of_iso639p1 lang1 in
    if lang1 = lang3 then None else Some lang3
  with Not_found -> None

let to_part2t_string lang =
  if Data.is_iso639p2t lang then Some (alpha3_of_int lang) else None

let to_part2b_string lang =
  if Data.is_iso639p2t lang
  then Some (alpha3_of_int (Data.to_iso639p2b lang)) else None

let of_part2_string s =
  try
    let lang = Data.of_iso639p2b (int_of_alpha3 s) in
    let lang3 = lang land 0x8000 in
    if Data.is_iso639p2t lang3 then Some lang3 else
    if Data.is_iso639p2t lang then Some lang else
    None
  with Not_found -> None

let of_part3_string s =
  try
    let lang = int_of_alpha3 s in
    if is_valid_part3 lang then Some lang else None
  with Not_found -> None

let of_part5_string s =
  try
    let lang = int_of_alpha3 s lor 0x8000 in
    if Data.is_iso639p5 lang then Some lang else
    None
  with Not_found -> None


lib/iso639.ml
(* Copyright (C) 2018  Petter A. Urkedal <paurkedal@gmail.com>
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version, with the OCaml static compilation exception.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *)

type t = int
let equal = (=)
let compare = compare
let to_int x = x

let is_valid lang =
  Char.code (Data.is_iso639p3_bits.[lang / 8]) lsr (lang mod 8) land 1 = 1

let of_int lang = if is_valid lang then Some lang else None

let of_int_exn lang =
  if is_valid lang then lang else invalid_arg "Iso639.of_int_exn"

let is_part1 x = Data.to_iso639p1 x <> x
let is_part2 x = Data.is_iso639p2t x
let is_part3 x = x < 0x8000
let is_part5 x = x > 0x8000

let alpha_of_int x = Char.chr (x mod 32 + 0x60)

let int_of_alpha = function
 | 'a'..'z' as c -> Char.code c - 0x60
 | _ -> raise Not_found

let to_alpha2 x = String.init 2 (fun i -> alpha_of_int (x lsr (5 * (1 - i))))
let to_alpha3 x = String.init 3 (fun i -> alpha_of_int (x lsr (5 * (2 - i))))

let to_part1_string lang =
  let lang1 = Data.to_iso639p1 lang in
  if lang1 = lang then None else Some (to_alpha2 lang1)

let to_part2t_string lang =
  if Data.is_iso639p2t lang then Some (to_alpha3 lang) else None

let to_part2b_string lang =
  if Data.is_iso639p2t lang
  then Some (to_alpha3 (Data.to_iso639p2b lang)) else None

let to_part3_string lang =
  if is_part3 lang then Some (to_alpha3 lang) else None

let to_part5_string lang =
  if is_part5 lang then Some (to_alpha3 lang) else None

let of_part1_string s =
  if String.length s <> 2 then None else
  try
    let lang1 = int_of_alpha s.[0] lsl 5 + int_of_alpha s.[1] in
    let lang3 = Data.of_iso639p1 lang1 in
    if lang1 = lang3 then None else Some lang3
  with Not_found -> None

let int_of_alpha3 s =
  if String.length s <> 3 then raise Not_found else
  int_of_alpha s.[0] lsl 10 + int_of_alpha s.[1] lsl 5 + int_of_alpha s.[2]

let of_part2_string s =
  try
    let lang = Data.of_iso639p2b (int_of_alpha3 s) in
    let lang3 = lang land 0x8000 in
    if Data.is_iso639p2t lang3 then Some lang3 else
    if Data.is_iso639p2t lang then Some lang else
    None
  with Not_found -> None

let of_part3_string s =
  try
    let lang = int_of_alpha3 s in
    if is_valid lang then Some lang else None
  with Not_found -> None

let of_part5_string s =
  try
    let lang = int_of_alpha3 s lor 0x8000 in
    if Data.is_iso639p5 lang then Some lang else
    None
  with Not_found -> None

let show lang =
  let s = to_alpha3 lang in
  if is_part3 lang then s else String.uppercase_ascii s

let pp ppf lang = Format.pp_print_string ppf (show lang)

let scope x = if is_part3 x then Data.lang3_scope x else `Collective


lib/lang_family.ml
(* Copyright (C) 2018  Petter A. Urkedal <paurkedal@gmail.com>
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version, with the OCaml static compilation exception.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *)

include Common

let of_int lang =
  if Data.is_iso639p5 lang then Some lang else None

let of_int_exn lang =
  if Data.is_iso639p5 lang then lang else
  invalid_arg "Iso639.Lang_family.of_int_exn"

let of_lang_or_family lang =
  let i = Lang_or_family.to_int lang in
  if i > 0x8000 then Some i else None

let to_lang_or_family = Lang_or_family.of_int_unsafe

let of_string = of_iso639p5


lib/lang_or_family.ml
(* Copyright (C) 2018  Petter A. Urkedal <paurkedal@gmail.com>
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version, with the OCaml static compilation exception.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *)

include Common

let of_int lang =
  if is_valid_part3 lang || Data.is_iso639p5 lang then Some lang else None

let of_int_exn lang =
  if is_valid_part3 lang || Data.is_iso639p5 lang then lang else
  invalid_arg "Iso639.of_int_exn"

let of_int_unsafe lang = lang

let is_part3 x = x < 0x8000
let is_part5 x = x >= 0x8000

let of_string s =
  try
    let lang = int_of_alpha3 s in
    if is_valid_part3 lang || Data.is_iso639p5 lang then Some lang else None
  with Not_found -> None

let to_part3_string lang =
  if is_part3 lang then Some (alpha3_of_int lang) else None

let to_part5_string lang =
  if is_part5 lang then Some (alpha3_of_int lang) else None

let scope x = if is_part3 x then Data.lang3_scope x else `Collective


lib/gen_data.ml
(* Copyright (C) 2018  Petter A. Urkedal <paurkedal@gmail.com>
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version, with the OCaml static compilation exception.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *)

open Printf

module Int_order = struct type t = int let compare = compare end
module Int_set = Set.Make (Int_order)
module Int_map = Map.Make (Int_order)

let invert_map m = Int_map.fold (fun k v -> Int_map.add v k) m Int_map.empty

module Tsv = struct

  let trim_bom s =
    let l = String.length s in
    if l = 0 || s.[0] <> '\xef' then s else begin
      assert (l >= 3 && s.[1] = '\xbb' && s.[2] = '\xbf');
      String.sub s 3 (l - 3)
    end

  let trim_cr s =
    let l = String.length s in
    if l > 0 && s.[l - 1] = '\r' then String.sub s 0 (l - 1) else s

  let read_row ?(sep = '\t') ic =
    let line = trim_cr (trim_bom (input_line ic)) in
    String.split_on_char sep line

  let iter ?sep ?header f fp =
    let ic = open_in fp in
    try
      (match header with
       | Some header ->
          let row0 = read_row ?sep ic in
          assert (row0 = header)
       | None -> ());
      while true do f (read_row ?sep ic) done
    with
     | End_of_file -> close_in ic
     | exn -> close_in ic; raise exn
end

let warn fmt = ksprintf (fun msg -> prerr_endline ("Warning: " ^ msg)) fmt

let gen_is_iso639p3_bits lang3_set =
  let is_iso639p3_bits = Bytes.create 4096 in
  lang3_set |> Int_set.iter begin function lang ->
    let i = lang / 8 in
    let bits = Char.code (Bytes.get is_iso639p3_bits i) in
    Bytes.set is_iso639p3_bits i (Char.chr (bits lor (1 lsl (lang mod 8))))
  end;
  printf "let is_iso639p3_bits = %S\n\n" (Bytes.to_string is_iso639p3_bits)

let gen_predicate fn lang_set =
  printf "let %s = function" fn;
  let index = ref 0 in
  lang_set |> Int_set.iter begin fun lang ->
    if !index mod 8 = 0 then printf "\n";
    printf " | %#x" lang;
    incr index;
  end;
  printf " -> true\n | _ -> false\n\n"

let gen_conversion fn m =
  printf "let %s = function\n" fn;
  Int_map.iter (fun k v -> printf " | %#x -> %#x\n" k v) m;
  printf " | lang -> lang\n\n"

let gen_scope_classifier fn m =
  printf "let %s = function\n" fn;
  let ctor_name = function
   | `Special -> "Special"
   | `Macro -> "Macro" in
  Int_map.iter (fun k v -> printf " | %#x -> `%s\n" k (ctor_name v)) m;
  printf " | _ -> `Individual\n\n"

let gen_one_to_many fn m =
  printf "let %s = function\n" fn;
  m |> Int_map.iter begin fun k v ->
    printf " | %#x -> \"" k;
    v |> List.iter (fun x -> printf "\\x%02x\\x%02x" (x lsr 8) (x land 0xff));
    printf "\"\n"
  end;
  print_string " | _ -> \"\"\n"

let alpha_of_int x = Char.chr (x mod 32 + 0x60)

let int_of_alpha = function
 | 'a'..'z' as c -> Char.code c - 0x60
 | _ -> raise Not_found

let int_of_alphaN s =
  let l = String.length s in
  assert (2 <= l && l <= 3);
  let rec loop i acc =
    if i = l then acc else
    loop (i + 1) (acc lsl 5 lor (int_of_alpha s.[i])) in
  loop 0 0

let int_option_of_alphaN = function "" -> None | s -> Some (int_of_alphaN s)

let alpha2_of_int x = String.init 2 (fun i -> alpha_of_int (x lsr (5*(1 - i))))
let alpha3_of_int x = String.init 3 (fun i -> alpha_of_int (x lsr (5*(2 - i))))

let merge_v3v2 ~what map_v3 map_v2 =
  let aux lang v3 v2 =
    (match v3, v2 with
     | None, None -> assert false
     | Some v3, Some v2 -> assert (v3 = v2); Some v3
     | Some v3, None ->
        warn "Code %S seen in ISO-639-3 to %s mapping \
              but not in the corresponding ISO-639-2 data."
             (alpha3_of_int lang) what;
        Some v3
     | None, Some v2 ->
        if lang < 0x8000 then
          warn "Non-collective language %S seen in ISO-639-2 to %s mapping \
                but not in the corresponding ISO-639-3 data."
               (alpha3_of_int lang) what;
        Some v2) in
  Int_map.merge aux map_v3 map_v2

let header3 = [
  "Id"; "Part2B"; "Part2T"; "Part1"; "Scope";
  "Language_Type"; "Ref_Name"; "Comment";
]
let header3m = ["M_Id"; "I_Id"; "I_Status"]
let header5 = [
  "URI"; "code"; "Label (English)"; "Label (French)";
]
let () =
  let lang3_set = ref Int_set.empty in
  let lang5_set = ref Int_set.empty in
  let lang2t_set = ref Int_set.empty in
  let lang_to_lang1_v2 = ref Int_map.empty in
  let lang_to_lang1_v3 = ref Int_map.empty in
  let lang_to_lang2b_v2 = ref Int_map.empty in
  let lang_to_lang2b_v3 = ref Int_map.empty in
  let lang3_scope_map = ref Int_map.empty in
  let lang3_macrolanguage_map = ref Int_map.empty in
  let lang3_macrolanguage_members_map = ref Int_map.empty in

  (* Load ISO-639-3 Data *)
  Sys.argv.(2) |> Tsv.iter ~header:header3 begin function
   | [lang3; lang2b; lang2t; lang1; scope; language_type; ref_name; comment] ->
      let lang3 = int_of_alphaN lang3 in
      lang3_set := Int_set.add lang3 !lang3_set;
      (match int_option_of_alphaN lang2t, int_option_of_alphaN lang2b with
       | None, None -> ()
       | Some lang2t, Some lang2b ->
          if lang2t <> lang2b then
            lang_to_lang2b_v3 := Int_map.add lang2t lang2b !lang_to_lang2b_v3;
          (match int_option_of_alphaN lang1 with
           | None -> ()
           | Some lang1 ->
              lang_to_lang1_v3 := Int_map.add lang2t lang1 !lang_to_lang1_v3)
       | _ -> assert false);
      (match scope with
       | "I" -> ()
       | "S" -> lang3_scope_map := Int_map.add lang3 `Special !lang3_scope_map
       | "M" -> lang3_scope_map := Int_map.add lang3 `Macro !lang3_scope_map
       | _ -> assert false)
   | _ -> assert false
  end;
  let lang3_scope lang =
    (match Int_map.find lang !lang3_scope_map with
     | exception Not_found -> `Individual
     | `Special -> `Special | `Macro -> `Macro)
  in
  Sys.argv.(3) |> Tsv.iter ~header:header3m begin function
   | [sM; sI; ("A" | "R")] ->
      let langM = int_of_alphaN sM in
      let langI = int_of_alphaN sI in
      assert (lang3_scope langM = `Macro);
      assert (lang3_scope langI = `Individual);
      assert (not (Int_map.mem langI !lang3_macrolanguage_map));
      lang3_macrolanguage_map :=
        Int_map.add langI langM !lang3_macrolanguage_map;
      let aux = function None -> Some [langI] | Some xs -> Some (langI :: xs) in
      lang3_macrolanguage_members_map :=
        Int_map.update langM aux !lang3_macrolanguage_members_map
   | _ -> assert false
  end;

  (* Load ISO-639-5 Data *)
  Sys.argv.(4) |> Tsv.iter ~header:header5 begin function
   | [_uri; part5; en; _fr] ->
      let lang = int_of_alphaN part5 lor 0x8000 in
      lang5_set := Int_set.add lang !lang5_set
   | _ -> assert false
  end;
  let lang5_set = !lang5_set in

  (* Load ISO-639-2 Data *)
  Sys.argv.(1) |> Tsv.iter ~sep:'|' begin function
   | ["qaa-qtz"; _; _; _; _] -> ()
   | [lang2b; lang2t; lang1; _label_en; _label_fr] ->
      let lang2b = int_of_alphaN lang2b in
      let lang2t = int_option_of_alphaN lang2t in
      let lang2t, lang2b =
        (match lang2t with
         | None -> (lang2b, None)
         | Some lang2t -> (lang2t, Some lang2b)) in
      let lang2t =
        if Int_set.mem (lang2t lor 0x8000) lang5_set
        then lang2t lor 0x8000 else lang2t in
      lang2t_set := Int_set.add lang2t !lang2t_set;
      (match int_option_of_alphaN lang1 with
       | Some lang1 ->
          lang_to_lang1_v2 := Int_map.add lang2t lang1 !lang_to_lang1_v2
       | None -> ());
      (match lang2b with
       | Some lang2b ->
          lang_to_lang2b_v2 := Int_map.add lang2t lang2b !lang_to_lang2b_v2
       | None -> ())
   | _ -> assert false
  end;

  let lang3_set = !lang3_set in
  let lang2t_set = !lang2t_set in
  let lang_to_lang1 =
    merge_v3v2 ~what:"ISO-639-1" !lang_to_lang1_v3 !lang_to_lang1_v2 in
  let lang_to_lang2b =
    merge_v3v2 ~what:"ISO-639-2B" !lang_to_lang2b_v3 !lang_to_lang2b_v2 in

  let lang_set = Int_set.union lang3_set lang5_set in
  Int_set.diff lang2t_set lang_set |> Int_set.iter begin fun lang ->
    warn "Exclunding %S from ISO-639-2 \
          as it is present in neither ISO-639-3 nor ISO-639-5 tables."
      (alpha3_of_int lang)
  end;

  gen_is_iso639p3_bits lang3_set;
  gen_predicate "is_iso639p5" lang5_set;
  gen_predicate "is_iso639p2t" (Int_set.inter lang2t_set lang_set);
  gen_conversion "to_iso639p2b" lang_to_lang2b;
  gen_conversion "of_iso639p2b" (invert_map lang_to_lang2b);
  gen_conversion "to_iso639p1" lang_to_lang1;
  gen_conversion "of_iso639p1" (invert_map lang_to_lang1);
  gen_scope_classifier "lang3_scope" !lang3_scope_map;
  gen_conversion "lang3_macrolanguage" !lang3_macrolanguage_map;
  gen_one_to_many "lang3_macrolanguage_members" !lang3_macrolanguage_members_map
